pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java

```diff
@@ -19,255 +19,409 @@
-      @ApiParam(value = "Name of segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
-      @ApiParam(value = "realtime|offline", required = false) @QueryParam("type") String tableTypeStr) {
+      @ApiParam(value = "Name of the segment", required = true) @PathParam("segmentName") @Encoded String segmentName,
+      @ApiParam(value = "OFFLINE|REALTIME") @QueryParam("type") String tableTypeStr) {
     segmentName = URIUtils.decode(segmentName);
-    CommonConstants.Helix.TableType tableType = Constants.validateTableType(tableTypeStr);
-    return listSegmentMetadataInternal(tableName, segmentName, tableType);
+    TableType tableType = Constants.validateTableType(tableTypeStr);
+    List<String> tableNamesWithType = getExistingTableNamesWithType(tableName, tableType);
+    List<Map<String, Object>> resultList = new ArrayList<>(tableNamesWithType.size());
+    for (String tableNameWithType : tableNamesWithType) {
+      Map<String, String> segmentMetadata = getSegmentMetadata(tableNameWithType, segmentName);
+      if (segmentMetadata != null) {
+        Map<String, Object> resultForTable = new LinkedHashMap<>();
+        resultForTable.put("tableName", tableNameWithType);
+        resultForTable.put("segmentMetadata", segmentMetadata);
+        resultList.add(resultForTable);
+      }
+    }
+    if (resultList.isEmpty()) {
+      String errorMessage =
+          "Failed to find segment: " + segmentName + " in table: " + tableName + (tableType != null ? (" of type: "
+              + tableType) : "");
+      throw new ControllerApplicationException(LOGGER, errorMessage, Status.NOT_FOUND);
+    }
+    return resultList;
   }
 
-  private String listSegmentMetadataInternal(@Nonnull String tableName, @Nonnull String segmentName,
-      @Nullable CommonConstants.Helix.TableType tableType) {
-    List<String> tableNamesWithType = getExistingTableNamesWithType(tableName, tableType);
+  @Nullable
+  private Map<String, String> getSegmentMetadata(String tableNameWithType, String segmentName) {
+    ZkHelixPropertyStore<ZNRecord> propertyStore = _pinotHelixResourceManager.getPropertyStore();
+    if (TableNameBuilder.OFFLINE.tableHasTypeSuffix(tableNameWithType)) {
+      OfflineSegmentZKMetadata offlineSegmentZKMetadata =
+          ZKMetadataProvider.getOfflineSegmentZKMetadata(propertyStore, tableNameWithType, segmentName);
+      return offlineSegmentZKMetadata != null ? offlineSegmentZKMetadata.toMap() : null;
+    } else {
+      RealtimeSegmentZKMetadata realtimeSegmentZKMetadata =
+          ZKMetadataProvider.getRealtimeSegmentZKMetadata(propertyStore, tableNameWithType, segmentName);
+      return realtimeSegmentZKMetadata != null ? realtimeSegmentZKMetadata.toMap() : null;
+    }
+  }
 
-    ArrayNode result = JsonUtils.newArrayNode();
+  @Deprecated
+  @GET
+  @Path("tables/{tableName}/segments/{segmentName}/metadata")
```

Comment:
Incompatible with existing API. Currently, this API returns a String. Please keep the api the same.